File: Milla-Rayne/android/app/src/main/java/com/millarayne/scene/SceneUtils.kt
/**
 * Scene utilities for Android
 * Provides scene generation and system detection
 */

package com.millarayne.scene

import android.content.Context
import android.content.res.Configuration
import android.os.PowerManager
import android.provider.Settings
import androidx.compose.ui.graphics.Color
import java.util.Calendar

/**
 * Get current time of day bucket
 */
fun getCurrentTimeOfDay(): TimeOfDay {
    val hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY)
    
    return when {
        hour in 5..7 -> TimeOfDay.DAWN
        hour in 8..16 -> TimeOfDay.DAY
        hour in 17..19 -> TimeOfDay.DUSK
        else -> TimeOfDay.NIGHT
    }
}

/**
 * Get color palette for time of day
 */
fun getPaletteForTimeOfDay(timeOfDay: TimeOfDay): ScenePalette {
    return when (timeOfDay) {
        TimeOfDay.DAWN -> ScenePalette(
            primary = 0xFFff9a8b,
            secondary = 0xFFffc3a0,
            accent = 0xFFffafbd,
            background = 0xFF1a1a2e
        )
        TimeOfDay.DAY -> ScenePalette(
            primary = 0xFF667eea,
            secondary = 0xFF764ba2,
            accent = 0xFFf093fb,
            background = 0xFF0f0f1e
        )
        TimeOfDay.DUSK -> ScenePalette(
            primary = 0xFFfa709a,
            secondary = 0xFFfee140,
            accent = 0xFFff8c00,
            background = 0xFF1e1e2e
        )
        TimeOfDay.NIGHT -> ScenePalette(
            primary = 0xFF4c669f,
            secondary = 0xFF3b5998,
            accent = 0xFF192f6a,
            background = 0xFF0a0a15
        )
    }
}

/**
 * Get accent color for app state
 */
fun getAccentForAppState(appState: AppState): Long {
    return when (appState) {
        AppState.IDLE -> 0xFF667eea
        AppState.LISTENING -> 0xFF10b981 // Green
        AppState.THINKING -> 0xFFf59e0b  // Amber
        AppState.SPEAKING -> 0xFF3b82f6  // Blue
    }
}

/**
 * Check if reduced motion is preferred
 * Checks both system animator duration scale and accessibility settings
 */
fun prefersReducedMotion(context: Context): Boolean {
    try {
        // Check animator duration scale (0 = animations disabled)
        val animatorScale = Settings.Global.getFloat(
            context.contentResolver,
            Settings.Global.ANIMATOR_DURATION_SCALE,
            1.0f
        )
        
        if (animatorScale == 0f) return true
        
        // Could also check transition animation scale
        val transitionScale = Settings.Global.getFloat(
            context.contentResolver,
            Settings.Global.TRANSITION_ANIMATION_SCALE,
            1.0f
        )
        
        return transitionScale == 0f
    } catch (e: Exception) {
        return false // Default to animations enabled if we can't check
    }
}

/**
 * Check if device is in power saver mode
 */
fun isInPowerSaverMode(context: Context): Boolean {
    return try {
        val powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager
        powerManager.isPowerSaveMode
    } catch (e: Exception) {
        false
    }
}

/**
 * Generate complete scene theme from context
 */
fun generateSceneTheme(
    timeOfDay: TimeOfDay,
    appState: AppState,
    reducedMotion: Boolean,
    performanceMode: PerformanceMode
): SceneTheme {
    val palette = getPaletteForTimeOfDay(timeOfDay).copy(
        accent = getAccentForAppState(appState)
    )
    
    // Animation speed based on reduced motion and performance mode
    val animationSpeed = when {
        reducedMotion -> 0f
        performanceMode == PerformanceMode.PERFORMANCE -> 0.5f
        performanceMode == PerformanceMode.HIGH_QUALITY -> 1f
        else -> 0.75f // BALANCED
    }
    
    // Parallax intensity based on performance mode
    val parallaxIntensity = when {
        reducedMotion -> 0f
        performanceMode == PerformanceMode.HIGH_QUALITY -> 1f
        performanceMode == PerformanceMode.BALANCED -> 0.5f
        else -> 0f // PERFORMANCE
    }
    
    return SceneTheme(
        palette = palette,
        gradientAngle = 135f,
        animationSpeed = animationSpeed,
        parallaxIntensity = parallaxIntensity
    )
}

File: Milla-Rayne/client/src/lib/scene/sceneUtils.ts
/**
 * Scene configuration and theming utilities
 * Provides asset-free, procedural scene generation based on context
 */

import type {
  TimeOfDay,
  AppState,
  SceneTheme,
  ScenePalette,
} from '@shared/sceneTypes';

/**
 * Get time of day bucket based on current hour
 */
export function getCurrentTimeOfDay(): TimeOfDay {
  const hour = new Date().getHours();

  if (hour >= 5 && hour < 8) return 'dawn';
  if (hour >= 8 && hour < 17) return 'day';
  if (hour >= 17 && hour < 20) return 'dusk';
  return 'night';
}

/**
 * Get color palette for a given time of day
 */
export function getPaletteForTimeOfDay(timeOfDay: TimeOfDay): ScenePalette {
  const palettes: Record<TimeOfDay, ScenePalette> = {
    dawn: {
      primary: '#ff9a8b',
      secondary: '#ffc3a0',
      accent: '#ffafbd',
      background: '#1a1a2e',
    },
    day: {
      primary: '#667eea',
      secondary: '#764ba2',
      accent: '#f093fb',
      background: '#0f0f1e',
    },
    dusk: {
      primary: '#fa709a',
      secondary: '#fee140',
      accent: '#ff8c00',
      background: '#1e1e2e',
    },
    night: {
      primary: '#4c669f',
      secondary: '#3b5998',
      accent: '#192f6a',
      background: '#0a0a15',
    },
  };

  return palettes[timeOfDay];
}

/**
 * Get accent color modifier for app state
 */
export function getAccentForAppState(appState: AppState): string {
  const accents: Record<AppState, string> = {
    idle: '#667eea',
    listening: '#10b981', // Green for listening
    thinking: '#f59e0b', // Amber for thinking
    speaking: '#3b82f6', // Blue for speaking
  };

  return accents[appState];
}

/**
 * Generate complete scene theme from context
 */
export function generateSceneTheme(
  timeOfDay: TimeOfDay,
  appState: AppState,
  reducedMotion: boolean,
  performanceMode: 'high-quality' | 'balanced' | 'performance'
): SceneTheme {
  const palette = getPaletteForTimeOfDay(timeOfDay);

  // Override accent based on app state
  palette.accent = getAccentForAppState(appState);

  // Animation speed based on reduced motion and performance mode
  let animationSpeed = 1;
  if (reducedMotion) {
    animationSpeed = 0;
  } else if (performanceMode === 'performance') {
    animationSpeed = 0.5;
  } else if (performanceMode === 'high-quality') {
    animationSpeed = 1;
  } else {
    animationSpeed = 0.75; // balanced
  }

  // Parallax intensity based on performance mode
  let parallaxIntensity = 0;
  if (!reducedMotion) {
    if (performanceMode === 'high-quality') {
      parallaxIntensity = 1;
    } else if (performanceMode === 'balanced') {
      parallaxIntensity = 0.5;
    }
    // performance mode has 0 parallax
  }

  return {
    palette,
    gradientAngle: 135, // Fixed diagonal gradient
    animationSpeed,
    parallaxIntensity,
  };
}

/**
 * Check if user prefers reduced motion
 */
export function prefersReducedMotion(): boolean {
  if (typeof window === 'undefined') return true;

  const mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)');
  return mediaQuery.matches;
}

/**
 * Check if page is backgrounded (for throttling animations)
 */
export function isPageBackgrounded(): boolean {
  if (typeof document === 'undefined') return false;
  return document.hidden;
}
